### PR TITLE
Add support for EssentialsDiscord

### PR DESCRIPTION
**Pull request by JRoy**
Adds support for EssentialsDiscord's `DiscordChatMessageEvent` in order to block chat messages from TownyChat channels which are not marked as `hooked: true` in their config. 

In other words, no message will be sent to the EssentialsDiscord `chat` channel unless users mark them as `hooked: true` in the towny channel config.

### CODE REVIEW

File: src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  public EssentialsDiscordHookListener(Chat plugin) {
+    this.plugin = plugin;
+  }
+
+  @EventHandler(ignoreCancelled = true)
+  public void onDiscordChat(DiscordChatMessageEvent event) {
+    Channel channel = plugin.getChannelsHandler().getChannel(event.getPlayer(), plugin.getTownyPlayerListener().directedChat.get(event.getPlayer()));
+    if (channel != null) {
+      event.setCancelled(!channel.isHooked());
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java:21]: Basing this off of whether a channel is hooked or not will break all of the other plugins using TownyChat, including DeluxeChat. A hooked channel means some other plugin is modifying the channel format.

File: src/com/palmergames/bukkit/TownyChat/listener/TownyChatPlayerListener.java [hunk 2]
```diff
@@ -116,22 +116,27 @@ public void onPlayerChat(AsyncPlayerChatEvent event) {
...
+						directedChat.remove(player);
 						return;
 					}
 					channel.chatProcess(event);
+					if (!Chat.usingEssentialsDiscord || event.isCancelled()) {
+						directedChat.remove(player);
+					}
 					return;
 				}
+				directedChat.remove(player);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/TownyChat/listener/TownyChatPlayerListener.java:139]: Not entirely sure why this change was made, directedChat.remove(player) should always occur, as this is when a player uses `/tc somechat` in order to speak into a channel which isn't the one they're currently speaking in.

File: src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java (hunk 1 above)
Comment (LlmDl) [reply src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java:21]: To elaborate: Many servers use DeluxeChat to alter their channels, including the global/general channel that everyone sees.

If we accepted this PR all of those established servers would find their general channel's chat cancelled.

I still am of the opinion that EssentialsDiscord is going to have to choose to be more deterministic on what it chooses to broadcast. We'll see what the other team members have to contribute.

File: src/com/palmergames/bukkit/TownyChat/channels/ChannelsHolder.java [hunk 3]
```diff
@@ -79,11 +79,13 @@ public Channel getChannel(String channelName) {
 	 * @return channel or null if not permitted
 	 */
 	public Channel getChannel(Player player, String command) {
-		
-		for (Channel channel: channels.values()) {
-			if (channel.getCommands().contains(command.toLowerCase())) {
-				if (TownyUniverse.getInstance().getPermissionSource().has(player, channel.getPermission()) || (channel.getPermission().isEmpty())) 
-					return channel;				
+
+		if (command != null) {
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/TownyChat/channels/ChannelsHolder.java:83]: Not entirely sure if this is needed, command should never be null.

**Review by LlmDl — CHANGES_REQUESTED** [review]
Upon looking this over again there is a flaw in how this was executed (using channel.isHooked())

File: src/com/palmergames/bukkit/TownyChat/listener/TownyChatPlayerListener.java (hunk 2 above)
Comment (JRoy) [reply src/com/palmergames/bukkit/TownyChat/listener/TownyChatPlayerListener.java:139]: This is done so `DiscordChatMessageEvent` can check to see if it was a direct channel, it will always end up being removed though, it's just done in `DiscordChatMessageEvent` if EssentialsDiscord was found.

File: src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  public void onDiscordChat(DiscordChatMessageEvent event) {
+    String directChat = plugin.getTownyPlayerListener().directedChat.get(event.getPlayer());
+    if (directChat != null) {
+      plugin.getTownyPlayerListener().directedChat.remove(event.getPlayer());
+      if (event.isCancelled()) {
+        return;
+      }
+      Channel channel = plugin.getChannelsHandler().getChannel(event.getPlayer(), directChat);
+      if (channel != null) {
+        event.setCancelled(!channel.isHooked() || channel.getType() != channelTypes.GLOBAL || channel.getRange() != -1);
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/TownyChat/listener/EssentialsDiscordHookListener.java:27]: I think the 3 `!channel.isHooked()` requirements can be dropped, as that would just lead to more unnecessary setup required by the EssDiscord users. Testing that its not global and not multi-world alone should be good.

**Review by LlmDl — CHANGES_REQUESTED** [review]

**Review by LlmDl — APPROVED** [review]
